File: javascript/easy/wordPattern.js
/**
 * @param {string} pattern
 * @param {string} str
 * @return {boolean}
 */
var wordPattern = function(pattern, str) {
    let map = new Map(), reverseMap = new Map();
    let words = str.split(" ");
    for(let i = 0; i < pattern.length; i++) {
        let c = pattern[i], word = words[i];
        if((map.get(c) !== undefined && map.get(c) !== word) ||
           (reverseMap.get(word) !== undefined && reverseMap.get(word) !== c)) {
            return false;
        }
        map.set(c, word);
        reverseMap.set(word, c);
    }
    return (pattern.length == words.length);
};

File: ruby/easy/word_pattern.rb
# @param {String} pattern
# @param {String} str
# @return {Boolean}
def word_pattern(pattern, str)
    map, reverse_map = {}, {}
    words = str.split(" ")
    i = 0
    pattern.each_char do |c|
        word = words[i]
        return false if (map[c] && map[c] != word) || (reverse_map[word] && reverse_map[word] != c)
        map[c] = word
        reverse_map[word] = c
        i += 1
    end
    return pattern.length == words.length
end
